common/beerocks/bcl/unit_tests/socket_event_loop_test.cpp

```diff
@@ -0,0 +1,196 @@
+/* SPDX-License-Identifier: BSD-2-Clause-Patent
+ *
+ * SPDX-FileCopyrightText: 2020 the prplMesh contributors (see AUTHORS.md)
+ *
+ * This code is subject to the terms of the BSD+Patent license.
+ * See LICENSE file for more details.
+ */
+
+#include <bcl/beerocks_socket_event_loop.h>
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+#include <easylogging++.h>
+
+// Initialize easylogging++
+INITIALIZE_EASYLOGGINGPP
+
+using namespace beerocks;
+
+using ::testing::InSequence;
+using ::testing::Invoke;
+using ::testing::StrictMock;
+
+static constexpr std::chrono::milliseconds s_test_timeout(100);
+
+/**
+ * @brief Mockable event handlers class
+ */
+using EventType = std::shared_ptr<Socket>;
+using LoopType  = EventLoop<EventType>;
+
+class EventHandlersMock : public LoopType::EventHandlers {
+
+public:
+    EventHandlersMock()
+    {
+        on_read    = [&](EventType socket, LoopType &loop) { return handle_read(socket, &loop); };
+        on_write   = [&](EventType socket, LoopType &loop) { return handle_write(socket, &loop); };
+        on_timeout = [&](EventType socket, LoopType &loop) {
+            return handle_timeout(socket, &loop);
+        };
+        on_disconnect = [&](EventType socket, LoopType &loop) {
+            return handle_disconnect(socket, &loop);
+        };
+        on_error = [&](EventType socket, LoopType &loop) { return handle_error(socket, &loop); };
+    }
+
+    MOCK_METHOD(bool, handle_read, ((EventType &), LoopType *));
+    MOCK_METHOD(bool, handle_write, ((EventType &), LoopType *));
+    MOCK_METHOD(bool, handle_timeout, ((EventType &), LoopType *));
+    MOCK_METHOD(bool, handle_disconnect, ((EventType &), LoopType *));
+    MOCK_METHOD(bool, handle_error, ((EventType &), LoopType *));
+};
+
+TEST(beerocks_socket_event_loop, setup)
+{
+    // Configure easylogging++ formatting
+    el::Configurations defaultConf;
+    defaultConf.setToDefault();
+    defaultConf.setGlobally(el::ConfigurationType::Format,
+                            "%level %datetime{%H:%m:%s:%g} <%thread> %fbase[%line] --> %msg");
+
+    // el::Loggers::addFlag(el::LoggingFlag::ColoredTerminalOutput);
+    el::Loggers::reconfigureLogger("default", defaultConf);
+}
+
+TEST(beerocks_socket_event_loop, simple_timeout)
+{
+    SocketEventLoop loop(s_test_timeout);
+    StrictMock<EventHandlersMock> timeout;
+
+    // Disable the on_write handler (to prevent the loop from firing "write ready" events)
+    timeout.on_write = nullptr;
+
+    // Create a dummy socket for checking the timeout mechanism
+    auto timeout_socket = std::make_shared<Socket>(socket(AF_UNIX, SOCK_DGRAM, 0));
+    ASSERT_NE(timeout_socket, nullptr);
```

Comment:
Excellent example of why this timeout interface sucks: in practical cases, you have to create a dummy socket to register a useful timeout :-)

Reply:
I still think that being able to set a timeout for sockets is useful (even if I don't have an example right now).
What if we'll also have separate `add_timeout` and `del_timeout` methods for setting arbitrary timeouts?